packages/node-runtime-worker-thread/src/serializer.ts

```diff
@@ -0,0 +1,145 @@
+  if (/^data:;base64,.+/.test(str)) {
+    return v8.deserialize(
+      Buffer.from(str.replace('data:;base64,', ''), 'base64')
+    );
+  }
+  return str;
+}
+
+function isPrimitive(
+  val: any
+): val is boolean | string | number | undefined | null {
+  return (typeof val !== 'object' && typeof val !== 'function') || val === null;
+}
+
+function isError(val: any): val is Error {
+  return val && val.name && val.message && val.stack;
+}
+
+function getNames<T>(obj: T): (keyof T)[] {
+  return Object.getOwnPropertyNames(obj) as (keyof T)[];
+}
+
+/**
+ * Extracts non-enumerable params from errors so they can be serialized and
+ * de-serialized when passed between threads.
+ *
+ * Even though v8 {de}serialize supports Error serialization, some data (e.g.,
+ * error `name` if you have a custom error) still can be lost during this
+ * process, so serializing it produces a better output.
+ */
+export function serializeError(err: Error) {
+  // Name is the only constructor property we care about
+  const keys = getNames(err).concat('name');
+  return keys.reduce((acc, key) => {
+    (acc as any)[key] = err[key];
+    return acc;
+  }, {} as Error);
+}
+
+/**
+ * Creates an instance of Error from error params (Error-like objects)
+ */
+export function deserializeError(err: any): Error {
+  return Object.assign(new Error(), err);
+}
+
+export enum SerializedResultTypes {
+  SerializedErrorResult = 'SerializedErrorResult',
+  InspectResult = 'InspectResult',
+  SerializedShellApiResult = 'SerializedShellApiResult'
+}
+
+type SerializedShellApiResult = {
+  origType: string;
+  serializedValue: Document;
+  nonEnumPropertyDescriptors: Record<string, PropertyDescriptor>;
+};
+
+export function serializeEvaluationResult(result: ShellResult): ShellResult {
+  result = { ...result };
+
+  // Type `null` indicates anything that is not returned by shell-api, any
+  // usual javascript value returned from evaluation
+  if (result.type === null) {
+    // Errors get special treatment to achieve better serialization
+    if (isError(result.rawValue)) {
+      result.type = SerializedResultTypes.SerializedErrorResult;
+      result.printable = serializeError(result.rawValue);
+    } else if (!isPrimitive(result.rawValue)) {
+      // For everything else that is not a primitive value there is just too
+      // many possible combinations of data types to distinguish between them,
+      // if we don't know the type and it's not an error, let's inspect and
+      // return an inspection result
+      result.type = SerializedResultTypes.InspectResult;
+      result.printable = inspect(result.rawValue);
+    }
+  } else if (!isPrimitive(result.printable)) {
+    // If type is present we are dealing with shell-api return value. In most
+    // cases printable values of shell-api are primitive or serializable as
+    // (E)JSON, but they also might have some non-enumerable stuff (looking at
+    // you, Cursor!) that we need to serialize additionally, otherwiser those
```

Comment:
I'm not sure a Cursor should not be serialised as is though, i'd like to understand how this works. Any `shell-api` value at this point should be just immediately serializable to EJSON without loss of important informations. Now this is probably necessary, but i'd feel better if we would have a stronger contract between this code and shell api and we could just relay on that fact.

Could you give an example on how not having this extra `nonEnumPropertyDescriptors ` would break the browser shell?

Reply:
`Cursor` uses non-enumerable `cursorHasMore` to decide if `Type "it" for more` should be printed or not:

https://github.com/mongodb-js/mongosh/blob/558f062e9614b87380fbb808017a1eaf4c69d826/packages/browser-repl/src/components/types/cursor-iteration-result-output.tsx#L24-L30

We could arguably make this property enumerable in shell-api itself, but this will break current `mongosh` output. CLI just inspects the object when printing it out, and if this prop is enumerable, it will be visible in the shell output:

https://github.com/mongodb-js/mongosh/blob/f11fe50a612ae2f383e3d90aa04ba0e1cdc12c55/packages/cli-repl/src/format-output.ts#L178-L188

I generally agree that making a stricter contract for `shell-api` might be beneficial in the long run, but would maybe suggest to handle this separately, what do you think?

Reply:
Decided f2f that we will change `shell-api` to avoid returning non-enumerable or non-serializable values from printable. Done in 38be23a